packages/angular_devkit/build_facade/src/architect.ts

```diff
@@ -0,0 +1,177 @@
+      `'${targetProperties.targetType}' target.`);
+  }
+}
+
+export class InvalidBuilderException extends BaseException {
+  constructor(targetProperties: TargetProperties) {
+    super(`Build facade '${targetProperties.buildFacade}' did not return a valid builder ` +
+      `for the following options: ${JSON.stringify(targetProperties, null, 2)}`);
+  }
+}
+
+export class InvalidRunnerException extends BaseException {
+  constructor(targetProperties: TargetProperties) {
+    super(`Build facade '${targetProperties.buildFacade}' did not return a valid runner ` +
+      `for the following options: ${JSON.stringify(targetProperties, null, 2)}`);
+  }
+}
+
+export interface TargetProperties {
+  root: Path;
+  projectType: string;
+  buildFacade: string;
+  targetType: string;
+}
+
+export interface TargetInfo<OptionsT = JsonObject> {
+  properties: TargetProperties;
+  options: OptionsT;
+}
+
+export class Architect {
+  _host: virtualFs.SyncDelegateHost<fs.Stats>;
+
+  constructor(private _workspace: Workspace) {
+    this._host = new virtualFs.SyncDelegateHost(new NodeJsSyncHost());
+  }
+
+  createTargetInfoFromWorkspace<OptionsT>(
+    projectName?: string,
+    targetName?: string,
+    configurationName?: string,
+    additionalInput?: Partial<OptionsT>,
+  ): TargetInfo<OptionsT> {
+
+    const project = projectName
+      ? this._workspace.getProjectByName(projectName)
+      : this._workspace.getDefaultProject();
+
+    if (!project) {
+      throw new ProjectNotFoundException(projectName);
+    }
+
+    const target = targetName
+      ? project.getTargetByName(targetName)
+      : project.getDefaultTarget();
+
+    if (!target) {
+      throw new TargetNotFoundException(targetName);
+    }
+
+    let configuration;
+
+    if (configurationName) {
+      configuration = project.getConfigurationByName(configurationName);
+
+      if (!configuration) {
+        throw new ConfigurationNotFoundException(configurationName);
+      }
+    }
+
+    // Resolve root for the target.
+    // TODO: add Path format to JSON schemas
+    const targetProperties: TargetProperties = {
+      root: resolve(this._workspace.path, normalize(project.root)),
+      projectType: project.projectType,
+      buildFacade: target.buildFacade,
+      targetType: target.targetType,
+    };
+
+    // tslint:disable-next-line:no-any
+    const targetOptions: { [k: string]: any } = {
+      ...target.props,
+      ...configuration ? configuration.props : {},
+      ...additionalInput ? additionalInput as {} : {},
+    };
+
+    // Remove unneeded properties.
+    delete targetOptions['buildFacade'];
+    delete targetOptions['targetType'];
+    delete targetOptions['default'];
+
+    return {
+      properties: targetProperties,
+      options: targetOptions as OptionsT,
+    };
+  }
+
+  // Will run the target (if the target supports run), using the targetInfo.
+  run<OptionsT>(
+    targetInfo: TargetInfo<OptionsT>, logger?: logging.Logger, stop?: Observable<void>,
```

Comment:
Is the `stop` observable a normal pattern? I thought a better (and more used) pattern was to `cancel()` the returned observable instead.

Reply:
I got it from https://medium.com/@benlesh/rxjs-dont-unsubscribe-6753ed4fda87. Might be dated though.

I don't see any `cancel()` on RxJs though. Do you mean `.unsubscribe()`? I don't think that allows for the cleanup, and that article advises against `.unsubscribe()`.